src/System.Private.ServiceModel/tests/Common/Infrastructure/src/CertificateManager.cs

```diff
@@ -0,0 +1,208 @@
+
+        // Retrieves a root certificate matching the given thumbprint from the root store
+        public static X509Certificate2 RootCertificateFromThumprint(string thumbprint)
+        {
+            return CertificateFromThumbprint(StoreName.Root, PlatformSpecificRootStoreLocation, thumbprint);
+        }
+
+        // Retrieves a root certificate matching the given issuer name and subject name from the root store
+        public static X509Certificate2 RootCertificateFromName(string issuerName, string subjectName)
+        {
+            return CertificateFromName(StoreName.Root, PlatformSpecificRootStoreLocation, issuerName, subjectName);
+        }
+
+        // Retrieves a client certificate matching the given thumbprint from the certificate store
+        public static X509Certificate2 ClientCertificateFromThumprint(string thumbprint)
+        {
+            return CertificateFromThumbprint(StoreName.My, StoreLocation.CurrentUser, thumbprint);
+        }
+
+        // Retrieves a client certificate matching the given issuer name and subject name from the certificate store
+        public static X509Certificate2 ClientCertificateFromName(string issuerName, string subjectName)
+        {
+            return CertificateFromName(StoreName.My, StoreLocation.CurrentUser, issuerName, subjectName);
+        }
+
+        private static X509Certificate2 CertificateFromName(StoreName storeName,
+                                                            StoreLocation storeLocation,
+                                                            string issuerName,
+                                                            string subjectName = null)
+        {
+            using (X509Store store = new X509Store(storeName, storeLocation))
+            {
+                try
+                {
+                    store.Open(OpenFlags.ReadOnly);
+                    X509Certificate2Collection certificates =
+                        store.Certificates.Find(X509FindType.FindByIssuerName, issuerName, validOnly: true);
+                    if (certificates.Count == 0)
+                    {
+                        return null;
+                    }
+
+                    if (subjectName != null)
+                    {
+                        certificates = certificates.Find(X509FindType.FindBySubjectName, subjectName, validOnly: true);
+                    }
+
+                    return certificates.Count == 0 ? null : certificates[0];
+                }
+                catch
+                {
+                    return null;
+                }
+            }
+        }
+
+
+        // Install the certificate into the Root store and returns its thumbprint.
+        // It will not install the certificate if it is already present in the store.
+        // It returns the thumbprint of the certificate, regardless whether it was added or found.
+        public static X509Certificate2 InstallCertificateToRootStore(X509Certificate2 certificate)
```

Comment:
Reviewers: mostly the same but removed all local caching the old code was doing.